presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SetFlatteningOptimizer.java

```diff
@@ -61,17 +63,35 @@ public PlanNode visitUnion(UnionNode node, RewriteContext<Boolean> context)
         {
             ImmutableList.Builder<PlanNode> flattenedSources = ImmutableList.builder();
             ImmutableListMultimap.Builder<Symbol, Symbol> flattenedSymbolMap = ImmutableListMultimap.builder();
+            flattenSetOperation(node, context, flattenedSources, flattenedSymbolMap);
+
+            return new UnionNode(node.getId(), flattenedSources.build(), flattenedSymbolMap.build(), ImmutableList.copyOf(flattenedSymbolMap.build().keySet()));
+        }
+
+        @Override
+        public PlanNode visitIntersect(IntersectNode node, RewriteContext<Boolean> context)
+        {
+            ImmutableList.Builder<PlanNode> flattenedSources = ImmutableList.builder();
+            ImmutableListMultimap.Builder<Symbol, Symbol> flattenedSymbolMap = ImmutableListMultimap.builder();
+            flattenSetOperation(node, context, flattenedSources, flattenedSymbolMap);
+
+            return new IntersectNode(node.getId(), flattenedSources.build(), flattenedSymbolMap.build(), ImmutableList.copyOf(flattenedSymbolMap.build().keySet()));
+        }
+
+        private void flattenSetOperation(SetOperationNode node, RewriteContext<Boolean> context, ImmutableList.Builder<PlanNode> flattenedSources, ImmutableListMultimap.Builder<Symbol, Symbol> flattenedSymbolMap)
+        {
             for (int i = 0; i < node.getSources().size(); i++) {
                 PlanNode subplan = node.getSources().get(i);
                 PlanNode rewrittenSource = context.rewrite(subplan, context.get());
 
-                if (rewrittenSource instanceof UnionNode) {
-                    // Absorb source's subplans if it is also a UnionNode
-                    UnionNode rewrittenUnion = (UnionNode) rewrittenSource;
-                    flattenedSources.addAll(rewrittenUnion.getSources());
+                Class<?> setOperationClass = node.getClass();
+                if (setOperationClass.isInstance(rewrittenSource)) {
+                    // Absorb source's subplans if it is also a SetOperation
```

Comment:
The comment would be more accurate if it said "... also a set operation of the same type"